### PR TITLE
add reconciliation mail sending

### PR DESCRIPTION
**Pull request by mdayican**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/BPS-1369

### Change description ###

add reconciliation mail sending

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationMailService.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+import java.util.Objects;
+import java.util.Optional;
+
+import static org.slf4j.LoggerFactory.getLogger;
+
+public class ReconciliationMailService {
+
+    private static final Logger logger = getLogger(ReconciliationMailService.class);
+
+    private final SupplierStatementRepository repository;
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationMailService.java:31]: there's more than one repository injected, let's name it correctly.

File: src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationMailService.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    private void processReconciliationReport(
+        ReconciliationReport reconciliationReport,
+        TargetStorageAccount account
+    ) {
+        try {
+            if (reconciliationReport.summaryContent != null) {
+
+                if (reconciliationReport.sentAt != null) {
+                    logger.info(
+                        "Skipping mailing process, Reconciliation Report send at {}, Report Id: {}",
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationMailService.java:95]: 'sent'

File: src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationMailService.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                    return;
+                }
+
+                SummaryReport summaryReport =
+                    objectMapper.readValue(reconciliationReport.summaryContent, SummaryReport.class);
+
+                File file = reconciliationCsvWriter.writeSummaryReconciliationToCsv(summaryReport);
+
+                Map<String, File> attachments = new HashMap<>();
+                attachments.put("Summary Report", file);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationMailService.java:108]: no file extension?

File: src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/task/ReconciliationMailTask.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+@Component
+@ConditionalOnProperty(value = "scheduling.task.send-reconciliation-report-mail.enabled")
+public class ReconciliationMailTask {
+
+    private static final String TASK_NAME = "send-reconciliation-report-mail";
+    private static final Logger logger = getLogger(ReconciliationMailTask.class);
+
+    private static final List<TargetStorageAccount> AVAILABLE_ACCOUNTS =
+        List.of(CFT, TargetStorageAccount.CRIME);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/task/ReconciliationMailTask.java:27]: let's be consistent

File: src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/task/ReconciliationMailTask.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    public ReconciliationMailTask(ReconciliationMailService reconciliationMailService) {
+        this.reconciliationMailService = reconciliationMailService;
+    }
+
+    @Scheduled(cron = "${scheduling.task.send-reconciliation-report-mail.cron}", zone = EUROPE_LONDON)
+    @SchedulerLock(name = TASK_NAME)
+    public void run() {
+        logger.info("Started {} job", TASK_NAME);
+
+        reconciliationMailService.process(LocalDate.now(UTC), AVAILABLE_ACCOUNTS);
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/task/ReconciliationMailTask.java:40]: things would be a bit simpler if `reconciliationMailService` took single account as parameter and iteration would happen here, but I guess it's a bit late for that ;)

Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/task/ReconciliationMailTask.java:40]: did not want to query same statement for each account.

File: src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationMailService.java (hunk 3 above)
Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationMailService.java]: prefix and suffix added.

File: src/test/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationMailServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+
+    @Mock
+    private ObjectMapper objectMapper;
+
+    private ReconciliationMailService service;
+
+    private static final String mailFrom = "from@f.com";
+    private static final String[] mailRecipients = {"r1@d.com"};
+    private static final List<TargetStorageAccount> AVAILABLE_ACCOUNTS =
+        List.of(CFT, TargetStorageAccount.CRIME);
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/reconciliation/service/ReconciliationMailServiceTest.java:72]: inconsistency

File: src/test/java/uk/gov/hmcts/reform/blobrouter/reconciliation/task/ReconciliationMailTaskTest.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+        // given
+        var reconciliationMailService = mock(ReconciliationMailService.class);
+        var task = new ReconciliationMailTask(reconciliationMailService);
+
+        // when
+        task.run();
+
+        // then
+        verify(reconciliationMailService, times(1))
+            .process(LocalDate.now(UTC), List.of(CFT, TargetStorageAccount.CRIME));;
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/blobrouter/reconciliation/task/ReconciliationMailTaskTest.java:29]: inconsistency

**Comment by apereverzin** [comment]
There should be an integration test for the scheduled task

**Comment by mdayican** [comment]
> There should be an integration test for the scheduled task

it would be same as test, I think it is fine without integration. we can consider to write functional tests

**Comment by apereverzin** [comment]
I mean `SchedulerConfigTest` for testing the scheduled configuration

**Comment by mdayican** [comment]
> I mean `SchedulerConfigTest` for testing the scheduled configuration

I think we do not need to test Spring.

**Comment by apereverzin** [comment]
We are testing our config, not Spring. We should either test all scheduled configs or not test any of them. As soon as all other scheduled configs are being tested it will be a minor effort to add a test for one more.

**Review by apereverzin — APPROVED** [review]

**Review by aliveni — APPROVED** [review]

**Pull request merged**